kernel/src/init.rs

```diff
@@ -0,0 +1,182 @@
+// In addition,
+// doing more work in the idle task may have negative impact on
+// the latency to switching from the idle task to a useful, runnable one.
+
+fn bsp_idle_loop() {
+    log::info!("[kernel] Idle thread for CPU #0 started");
+
+    // Spawn the first non-idle kernel thread on BSP.
+    ThreadOptions::new(first_kthread)
+        .cpu_affinity(CpuId::bsp().into())
+        .sched_policy(SchedPolicy::default())
+        .spawn();
+
+    // Wait till the init process is spawned.
+    let init_process = loop {
+        if let Some(init_process) = INIT_PROCESS.get() {
+            break init_process;
+        };
+
+        ostd::task::halt_cpu();
+    };
+
+    // Wait till the init process becomes zombie.
+    while !init_process.status().is_zombie() {
+        ostd::task::halt_cpu();
+    }
+
+    // TODO: exit via qemu isa debug device should not be the only way.
+    let exit_code = if init_process.status().exit_code() == 0 {
+        QemuExitCode::Success
+    } else {
+        QemuExitCode::Failed
+    };
+    exit_qemu(exit_code);
+}
+
+fn ap_idle_loop() {
+    log::info!(
+        "[kernel] Idle thread for CPU #{} started",
+        // No races because this function runs on a certain AP.
+        CpuId::current_racy().as_usize(),
+    );
+
+    loop {
+        ostd::task::halt_cpu();
+    }
+}
+
+//--------------------------------------------------------------------------
+// The first kernel thread
+//--------------------------------------------------------------------------
+
+// The main function of the first (non-idle) kernel thread
+fn first_kthread() {
+    println!("[kernel] Spawn the first kernel thread");
+
+    let init_mnt_ns = MountNamespace::get_init_singleton();
+    let fs_resolver = init_mnt_ns.new_fs_resolver();
+    init_in_first_kthread(&fs_resolver);
+
+    print_banner();
+
+    INIT_PROCESS.call_once(|| {
+        let karg: KCmdlineArg = boot_info().kernel_cmdline.as_str().into();
+        spawn_init_process(
+            karg.get_initproc_path().unwrap(),
+            karg.get_initproc_argv().to_vec(),
+            karg.get_initproc_envp().to_vec(),
+        )
+        .expect("Run init process failed.")
+    });
+}
+
+static INIT_PROCESS: Once<Arc<Process>> = Once::new();
+
+fn init_in_first_kthread(fs_resolver: &FsResolver) {
+    component::init_all(InitStage::Kthread, component::parse_metadata!()).unwrap();
+    // Work queue should be initialized before interrupt is enabled,
+    // in case any irq handler uses work queue as bottom half
+    crate::thread::work_queue::init_in_first_kthread();
+    crate::net::init_in_first_kthread();
+    crate::fs::init_in_first_kthread(fs_resolver);
+    crate::ipc::init_in_first_kthread();
+    #[cfg(any(target_arch = "x86_64", target_arch = "riscv64"))]
+    crate::vdso::init_in_first_kthread();
+}
+
+fn print_banner() {
+    println!("");
+    println!("{}", logo_ascii_art::get_gradient_color_version());
+}
+
+pub(super) fn init_in_first_process(ctx: &Context) {
```

Comment:
Rename this method to `on_first_process_startup` to convey the intention that this function is a callback.

The visibility needs to be changed to `pub(crate)` as `lib.rs` won't import this symbol.